lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -332,20 +332,24 @@ hoistSpecialInstruction(std::unique_ptr<LoopNestSummary> &LoopSummary,
   bool Changed = false;
 
   for (auto *Inst : Special) {
-    auto *BI = dyn_cast<BeginAccessInst>(Inst);
-    assert(BI && "Only BeginAccessInst are supported");
-    SmallVector<EndAccessInst *, 2> Ends;
-    getEndAccesses(BI, Ends);
-    if (!hoistInstruction(DT, BI, Loop, Preheader)) {
+    if (!hoistInstruction(DT, Inst, Loop, Preheader)) {
       continue;
     }
-    LLVM_DEBUG(llvm::dbgs() << "Hoisted " << *BI);
-    for (auto *instSink : Ends) {
-      if (!sinkInstruction(DT, LoopSummary, instSink, LI)) {
-        llvm_unreachable("LICM: Could not perform must-sink instruction");
+    if (auto *BI = dyn_cast<BeginAccessInst>(Inst)) {
+      SmallVector<EndAccessInst *, 2> Ends;
+      getEndAccesses(BI, Ends);
+      LLVM_DEBUG(llvm::dbgs() << "Hoisted BeginAccess " << *BI);
+      for (auto *instSink : Ends) {
+        if (!sinkInstruction(DT, LoopSummary, instSink, LI)) {
+          llvm_unreachable("LICM: Could not perform must-sink instruction");
```

Comment:
That looks pretty scary. Can you prove that whenever hoistInstruction succeeds also sinkInstruction will succeed?
The logic in both functions is very different. Just a simple example: what about an infinite loop? Wouldn't hoisting succeed but sinking not?

Is there a way to bail if on of hoisting/sinking fails? Just to be on the safe side?

Reply:
There's no way to bail, but, all the appropriate checks where done when analyzing the begin access instruction. as long as the loop is contains at least one exiting block this should succeed. `sinkInstruction` can only fail in the non-special-instructions case. just re-using the code.

Reply:
ok, as long as you are sure about it :-)